pkg/encoding/kzgEncoder/multiframe.go

```diff
@@ -0,0 +1,171 @@
+package kzgEncoder
+
+import (
+	"bytes"
+	"encoding/gob"
+	"errors"
+	"fmt"
+
+	rs "github.com/Layr-Labs/eigenda/pkg/encoding/encoder"
+	bls "github.com/Layr-Labs/eigenda/pkg/kzg/bn254"
+)
+
+// Sample is the basic unit for a verification
+// A blob may contain multiple Samples
+type Sample struct {
+	Commitment bls.G1Point
+	Proof      bls.G1Point
+	Row        int // corresponds to a row in the verification matrix
+	Coeffs     []bls.Fr
+	X          uint // X is assignment
+}
+
+// generate a random value using Fiat Shamir transform
+// we can also pseudo randomness generated locally, but we have to ensure no adv can manipulate it
+// Hashing everything takes about 1ms, so Fiat Shamir transform does not incur much cost
+func GenRandomness(params rs.EncodingParams, samples []Sample, m int) (bls.Fr, error) {
+	var buffer bytes.Buffer
+	enc := gob.NewEncoder(&buffer)
+
+	for _, sample := range samples {
+		err := enc.Encode(sample.Commitment)
+		if err != nil {
+			return bls.ZERO, err
+		}
+	}
+
+	var randomFr bls.Fr
+
+	err := bls.HashToSingleField(&randomFr, buffer.Bytes())
+	if err != nil {
+		return bls.ZERO, err
+	}
+
+	return randomFr, nil
+}
+
+// UniversalVerify implements batch verification on a set of chunks given the same chunk dimension (chunkLen, numChunk).
```

Comment:
The chunkLen will depend on blobLen, so if the blob sizes in a batch are spread widely, that seems to diminish the benefit of universal verification, any perf test/analysis for that?

Reply:
The blobLen can only be power of 2. So we will have [2,4,8, 16,32,64,128,256,512,1024...], up some limit. Given a large quantization factor and number of nodes, the chunk length is smaller value. So the set of possible blob Length is smaller.

The benchmark test above establish the baseline which allows latency calculation. Suppose you have a m subBatch each with d_i blob. 

You can compute the total latency as max_i(d_i/100*100 ms), the first 100 is for 100 blobs, the second 100 is for 100ms (see the [link](https://docs.google.com/spreadsheets/d/1ilkYxcG3ktEFOA0poUgH1CkUaZy12Cp2dvEiQ4NU-jM/edit#gid=574424216)). That 100ms can be further reduced because we will parallelize the length proof verification. We take max over m subBatch, because subBatch verification can be parallelize. If you have limited number of thread k. Then the total time is throughput (max_i(d_i/100*100 ms))*m/k

Reply:
Sounds fair, it looks the cardinality of <chunkLen, numChunks> is limited due to the exponential growth of values, so the worst case there may just about ~10 sub batches